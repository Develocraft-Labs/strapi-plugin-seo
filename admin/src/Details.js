import React, { memo, useCallback } from "react";
import { useParams, useHistory } from "react-router";
import { LoadingIndicatorPage } from "strapi-helper-plugin";
import Preview from "./Preview";

import BackArrow from "./ui/Svgs/Back";
import styled from "styled-components";
import { Column } from "./components/ui/common";
import saveSeoToItem from "./utils/saveSeoToItem";
import { deleteSeo } from "./api/seoApi";
import useResource from "./hooks/useResource";
import useSeoDetails from "./hooks/useSeoDetails";
import pluginId from "./pluginId";

const BackButton = styled.button`
  padding: 0.5rem 0rem;
  width: fit-content;
  margin-bottom: 2rem;
`;

const DetailsContainer = styled(Column)`
  position: relative;
  padding: 0px 30px 66px;
  max-width: 100%;
`;

const Details = () => {
  const { uid, locale, seoName, collectionTypeId, id } = useParams();
  const history = useHistory();
  const resource = useResource(uid, collectionTypeId);
  const { isLoading } = resource;

  const selectedSeo = useSeoDetails({ locale, seoName });
  const handleBackButton = () => {
    history.push("/plugins/seo");
  };
  const handleSave = useCallback((seo) => {
    selectedSeo.setSeo(seo);
    history.replace(
      `/plugins/${pluginId}/${uid}/details/${id}/${locale}/${seo.seoName}/${collectionTypeId}`
    );
  });

  const handleDeleteSeo = async ({ id }) => {
    const response = await deleteSeo({ id });

    if (response) {
      const seo = null;
      const id = +collectionTypeId;

      await saveSeoToItem({ seo, uid, id });

      handleBackButton();
    }
  };
  if (isLoading || selectedSeo.isLoading) {
    return <LoadingIndicatorPage />;
  }

  return (
    <DetailsContainer fullWidth>
      <BackButton onClick={handleBackButton}>
        <BackArrow />
      </BackButton>
      <Preview
        locale={locale}
        selectedSeo={selectedSeo.seo}
        resource={resource}
        onSave={handleSave}
        handleDeleteSeo={handleDeleteSeo}
        collectionTypeId={collectionTypeId}
      />
    </DetailsContainer>
  );
};

export default memo(Details);
