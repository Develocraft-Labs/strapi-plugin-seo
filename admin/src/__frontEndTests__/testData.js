import trads from "../translations";

export const id = 1,
  firstname = "John",
  lastname = "Doe",
  email = "John.Doe@companyA.com",
  title = "title",
  created_by = {
    id,
    email,
    firstname,
    lastname,
  },
  created_at = "2021-10-12T11:49:01.042Z",
  updated_at = "2021-10-15T09:43:36.344Z",
  updated_by = { email, firstname, id, isActive: true, lastname },
  seo = {
    canonical: "https://www.companyA.com/articles",
    created_at: "2021-10-12T11:55:00.595Z",
    created_by: 1,
    id: 1,
    jsonLd: {},
    metaDescription: "Artists Desc",
    metaTitle: "Artists Title",
    ogImage: "Artist og image",
    ogTitle: "Artists Og Title",
    ogType: "Artists Og Type",
    ogUrl: "Artists og url",
    title: "Artists Seo",
    updated_at,
    updated_by: 1,
  },
  resourceData = {
    title,
    id,
    created_by,
  },
  collectionName = "Artists",
  pagination = { page: 1, total: 13, pageSize: 5, pageCount: 3 },
  uid = "application::artists.artists",
  fullResults = [
    { created_at, created_by, id, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 2, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 3, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 4, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 5, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 6, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 7, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 8, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 9, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 10, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 11, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 12, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 13, seo, title, updated_at, updated_by },
  ],
  seoId = 1,
  apiID = "artists",
  attributes = {
    seo: {
      dominant: false,
      nature: "oneWay",
      plugin: "seo",
      private: false,
      target: "plugins::seo.seo",
      unique: false,
    },
    title,
  },
  schema = {
    attributes,
    collectionName,
    connection: "default",
    description: "",
    draftAndPublish: true,
    kind: "collectionType",
    name: "Artists",
    pluginOptions: {},
    restrictRelationsTo: null,
    visible: true,
  },
  resourceSchema = { collectionName, schema },
  pluginUID = "plugins::seo.seo",
  projectCollectionTypes = [{ apiID, schema, uid }],
  selectedSeo = {
    canonical: "https://www.companyA.com/articles",
    created_at: "2021-10-12T11:55:00.595Z",
    created_by: created_by,
    id: 1,
    jsonLd: {},
    metaDescription: "Artists Desc",
    metaTitle: "Artists Title",
    ogImage: "Artist og image",
    ogTitle: "Artists Og Title",
    ogType: "Artists Og Type",
    ogUrl: "Artists og url",
    title: "Artists Seo",
    updated_at,
    updated_by: updated_by,
  },
  refreshOptions = () => {},
  metaTitle = "meta title",
  metaDescription = "meta description",
  canonical = "canonical",
  handleDeleteSeo = () => {},
  handleEditSeo = () => {},
  defaultLocale = "en",
  userEnabledLocales = [
    {
      code: "en",
      created_at: "2021-12-20T09:03:53.724Z",
      id: 1,
      isDefault: true,
      name: "English (en)",
      updated_at: "2021-12-20T09:03:53.724Z",
    },
  ],
  seos = [seo, fullResults];

export const resourceComponentTestData = {
  resourceData,
  resourceSchema,
};

export const seoFormComponentTestData = {
  pluginUID,
  seoId,
  projectCollectionTypes,
  resource: resourceComponentTestData,
};

export const contentTypesListComponentTestData = [
  {
    collectionName,
    fullResults,
    pagination,
    uid,
  },
];

export const translationPickerComponentTestData = {
  translations: trads,
  handleTranslation: () => {},
  value: "",
};

export const contentTypesItemComponentTestData = {
  uid,
  fullResults,
  pagination,
  collectionName,
};
export const previewComponentTestData = {
  selectedSeo,
  pluginUID,
  refreshOptions,
  projectCollectionTypes,
  handleDeleteSeo,
};

export const tableComponentTestData = {
  seos: { collectionName, fullResults, uid, pagination },
  handleDeleteSeo,
  handleEditSeo,
  userEnabledLocales,
  uid,
  defaultLocale,
};
