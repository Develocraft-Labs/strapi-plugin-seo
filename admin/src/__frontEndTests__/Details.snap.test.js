/**
 * @jest-environment jsdom
 */
import React from "react";
import renderer from "react-test-renderer";

import Details from "../Details";
import { seoFormComponentTestData } from "./testData";

// mocking components from strapi-helper-plugin
jest.mock("strapi-helper-plugin", () => {
  const Loader = jest.fn(({ children }) => children);
  const LoadingIndicator = jest.fn((props) => (
    <Loader
      onChangeParams={props.handleParamsChange}
      params={props.params}
      count={props.count}
    >
      <div>loading...</div>
    </Loader>
  ));

  const request = jest.fn().mockImplementationOnce(() => {});

  const Input = jest.fn((props) => (
    <input value={props.value} onChange={props.onChange} name={props.name} />
  ));
  const InputText = jest.fn((props) => <Input props={props} />);

  const options = (props) => (
    <option value={props.title} id={props.id}>
      {props.title}
    </option>
  );
  const ReactSelect = jest.fn((props) => <Input props={props} />);
  const Select = jest.fn((props) => (
    <ReactSelect selectedValue={props.value} onChange={props.handleChange}>
      {options}
    </ReactSelect>
  ));

  const TextArea = jest.fn((props) => <Input props={props} />);
  const InputTextArea = jest.fn((props) => (
    <TextArea value={props.value} onChange={props.onChange} name={props.name} />
  ));

  const JsonLd = jest.fn((props) => <Input props={props} />);
  const InputJson = jest.fn((props) => <Input props={props} />);
  const InputsIndex = jest.fn((props) => (
    <JsonLd
      type="json"
      value={props.value}
      onChange={props.onChange}
      name={props.name}
      customInputs={{ json: InputJson }}
    />
  ));
  return {
    LoadingIndicatorPage: LoadingIndicator,
    InputText: InputText,
    InputTextArea: InputTextArea,
    InputsIndex: InputsIndex,
    request: request,
    Select: Select,
  };
});

// mock useParams from react-router
jest.mock("react-router", () => ({
  useParams: jest.fn().mockReturnValue({
    id: 1,
    uid: "application::artists.artists",
  }),
  useHistory: jest.fn().mockReturnValue([]),
}));

describe("SNAP TESTING DETAILS COMPONENT", () => {
  const component = renderer.create(
    <Details
      pluginUID={seoFormComponentTestData.pluginUID}
      seoId={seoFormComponentTestData.seoId}
      collectionTypes={seoFormComponentTestData.projectCollectionTypes}
    />
  );
  const tree = component.toJSON();
  it("Should Render Successfully", () => {
    expect(component).toBeDefined();
    expect(tree).toMatchSnapshot();
  });
});
