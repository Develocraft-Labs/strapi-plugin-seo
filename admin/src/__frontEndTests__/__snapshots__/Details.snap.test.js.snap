// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAP TESTING DETAILS COMPONENT Should Render Successfully 1`] = `
<div>
  loading...
</div>
`;
